### ISSUE TITLE
Provide option to configure format of branch name

### ISSUE THREAD

**Issue by greysteil** [dependabot/dependabot-core#396]
_From @edkellena on January 9, 2018 15:40_

Hey,
It would be super awesome if we could specify some format for the branch name created.

Use-case:
Our CI system sees a PR raised and will run through a build process that generates a Docker image and tags it with the branch name.  Unfortunately Docker tags have more limited scope than Git branches...
`A tag name must be valid ASCII and may contain lowercase and uppercase letters, digits, underscores, periods and dashes. A tag name may not start with a period or a dash and may contain a maximum of 128 characters.`

The `/` in the Dependabot branches is causing that tagging to fail.  We can work around this with some awk/shell wizardry I guess, but would be nice not to have to do this.

_Copied from original issue: dependabot/feedback#72_

**Comment by greysteil** [comment]
Thanks for the feedback @edkellena.

It's a totally legit request, but I'm not keen on changing things on Dependabot's side for a couple of reasons:
- branch name generation is already [quite complicated](https://github.com/dependabot/dependabot-core/blob/master/lib/dependabot/pull_request_creator/branch_namer.rb) and I don't want to add more considerations to it
- there's no obvious place to put the setting in the frontend at the moment

Do you have control over the build process on your side? I'm hoping it would be straightforward for you to sanitize the branch name when generating a tag.

**Comment by greysteil** [comment]
_From @edkellena on January 9, 2018 22:12_

Fair enough 👍 I did wonder whether it would be a fruitless ask!   Would be happy to attempt a PR for the generation, but without the UI setting available, that would be a bit pointless.

I do have some very hacky workaround at the moment, which I'll try to unhack soon.  Cheers though!

**Comment by greysteil** [comment]
Always worth asking! Thanks for understanding 🙂 .

**Comment by SanjiKir** [comment]
Is it still a problem to implement? We have exactly the same issue

**Comment by BobbyMcWho** [comment]
You could always monkey-patch it, now whether or not you _should_... 
```ruby
module Dependabot
  class PullRequestCreator
    class BranchNamer
      def new_branch_name
        super.gsub(TO_REMOVE, TO_REPLACE)
      end
    end
  end
end
```

**Comment by ritaly** [comment]
if you use dependabot v2  just add to your `dependabot.yml` 

```
pull-request-branch-name:
      separator: "-"
```

https://docs.github.com/en/github/administering-a-repository/configuration-options-for-dependency-updates#pull-request-branch-nameseparator

**Comment by pavellishin** [comment]
I'd like to bump this request - our CI system also requires that branch names only be so long.

What we _don't_ require is that our branch names be semantically accurate - something like `dependabot_20210625.1544` would work just as well as a descriptive name, given that the pull request title and description, as well as the contents, are enough to go on.

I do understand that the logic in `branch_namer` is moderately complex, and it would be hard to set up some sort of way to provide a format string, but maybe a flag to toggle the _type_ of branch name format? Or even an option to leave off certain parts (e.g., package manager) and/or truncate the whole thing (from the start or the end) to a given length?

Right now, a lot of our dependabot PRs are unusable without a dev having to manually create a new branch and PR.

**Comment by oliviasv** [comment]
Is there an option to change the branch name? I think it is too long, and maybe it would be better and cleaner for the branch to include only dependabot/updated-dependencies.

**Comment by jayoutdoorsy** [comment]
Why was this closed? The following dependabot branch breaks docker tags `dependabot/npm_and_yarn/shelljs-0.8.5`. I feel like the obvious fix would be to switch the branch naming convention to `dependabot/npm_and_yarn/${lib_name}-${uuid.slice(0, 8)}`. We will be forced to do this in our own code. There's already an option in the yaml file for `pull_request_branch_name` but the only sub-option is `separator`

**Comment by luiscantero** [comment]
+1 for this request. Our dependabot build fail because we use ACR.
ACR requirements: Repository names can only include lowercase alphanumeric characters, periods, dashes, underscores, and forward slashes. [Reference](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-concepts#repository)

Sample error from the pipeline:
```
Message: Image name 'dependabot/nuget/Moq-4.1/something/something:2.2.0-linux-amd64' is invalid. 
The acceptable patterns are 'repository' or 'repository:tag'.
The repository and tag names follow the standardized docker repository and tag naming conventions.
Tag templates are supported within '{{' and '}}'. 
```

**Comment by jeffwidman** [comment]
I'm going to re-open since there's so much interest here. I'm not sure that we'll choose to solve this by providing the option, but clearly there's a problem that needs solving in one way or the other.

**Comment by dcsaszar** [comment]
FWIW Another data point:

Our CI end-to-end tests check PRs (including dependabot PRs) through Cloudflare preview deploys. At some point, for branch names longer than ~28 characters, Cloudflare adds an (undocumented 🙈) 4-random-letter suffix to the [otherwise deterministic preview URL](https://developers.cloudflare.com/pages/platform/preview-deployments/#preview-aliases). This makes the CI fail dependabot PRs with long branch names as it can't easily compute the preview URL by itself.

For us, a `dependabot.yml` option to limit the dependabot branch name length (e.g. to 28) would be a great start.

**Comment by Spazzy757** [comment]
Hey just to add a different usecase, we have a CI/CD that deploys ephemeral environments for branches and runs E2E tests on them. The branch name sepearator replaces the `/` with no issue but our deployment does not support `.` (Think of the case of using the branch name in a URL, it creates a TLS issue when you use that many `.`).

**Comment by deivid-rodriguez** [comment]
Hi!

If I understand correctly, there’s two issues reported here:

* Branch name using not acceptable characters as separators.
* Branch name being too long.

The former is already fixed by the `pull_request_branch_name.separator` option. For the latter, truncating the max branch name length through a `pull_request_branch_name.max_length` configuration seems like the easiest way, specially because [we already have pull request creation logic in place](https://github.com/dependabot/dependabot-core/blob/42425eb2f8f49f12d34c400c8bc18ab6789c033b/common/lib/dependabot/pull_request_creator/branch_namer.rb#L20) to do that.

This will require a few internal changes but it seems like a big part of it is already in place.

**Comment by luiscantero** [comment]
@deivid-rodriguez The other issue is the use of [uppercase chars](https://github.com/dependabot/dependabot-core/issues/396#issuecomment-1055239459).

**Comment by deivid-rodriguez** [comment]
Oh I see, thanks for noticing that.

**Comment by dangerground** [comment]
The branch-length problem also regularly hits us, when using [GCP Workload Identity Federation](https://cloud.google.com/blog/products/identity-security/enabling-keyless-authentication-from-github-actions) with the following error:

`
google-github-actions/auth failed with: retry function failed after 1 attempt: failed to generate Google Cloud federated token for projects/XXX/locations/global/workloadIdentityPools/YYY/providers/ZZZ: (400) {"error":"invalid_request","error_description":"The size of mapped attribute google.subject exceeds the 127 bytes limit. Either modify your attribute mapping or the incoming assertion to produce a mapped attribute that is less than 127 bytes."}
`

This emerges from the [Token-Format](https://github.com/google-github-actions/auth#github-token-format) where the subject is the repository- and branch-name, which in combination with the [length limitation](https://cloud.google.com/iam/docs/configuring-workload-identity-federation#mappings-and-conditions) on GCP for the subject.

The only workaround we currently have is to rename the branch and create a new PR, because the old one is tied to the branch name, so that our checks can work as expected.

Providing/Exposing the `max-length` configuration for `depenabot.yml` really would help us.

**Comment by ramachandran-murugaian-kr** [comment]
I am also facing the same issues with branch name length. We are using harness to deploy our applications and it is allowing only 63 characters for the branch name.

 It would be great if we get the option for max-length

**Comment by TomNaessens** [comment]
The option to configure maximum branch length has been released here in v0.213.0 (https://github.com/dependabot/dependabot-core/blob/main/CHANGELOG.md#v02130-31-october-2022) and in dependabot-gitlab in version v.0.30.0 (https://gitlab.com/dependabot-gitlab/dependabot/-/releases/v0.30.0).

**Comment by ramachandran-murugaian-kr** [comment]
@TomNaessens Thank you for the update. How do I specific this specific version in depdendabot.yml. I am receiving below error 

`The property '#/updates/0/pull-request-branch-name' contains additional properties ["max_length"] outside of the schema when none are allowed
The property '#/updates/1/pull-request-branch-name' contains additional properties ["max_length"] outside of the schema when none are allowed`

**Comment by deivid-rodriguez** [comment]
I _think_ the problem is that while the option was added to this library, we're not yet making use of it in the GitHub service, nor have yet exposed any configuration option for it, since that requires a number of internal changes. But this feature is definitely closer now than it was before!

**Comment by dcsaszar** [comment]
@ramachandran-murugaian-kr @deivid-rodriguez I agree.

To my understanding what @TomNaessens referred to is that the internal branch namer now has the _ability_ to generate branch names of limited length (`branch_name_max_length`). Maybe Tom was implying that a quick solution for this issue would be to _expose and document_ this option to the YAML config (maybe as `pull-request-branch-name.max-length`)?

**Comment by lascufa** [comment]
We are also looking forward to a feature which enables the customization of the dependabot branch name. The issue we have is that we create a lot of resources based on the branch name. Our dependabot builds fail, because resource names (e.g. included in domain names, in AWS resources, ...) are reaching the allowed byte limits.

For example a branch name like `dependabot-[unique ID]` where unique ID could e.g. be just an epoch timestamp or something like it would be awesome.

**Comment by gczobel** [comment]
In our case, we are running some k3s infrastructure and we are limited to branch names of 63 characters
https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set

Exposing the max-length will be great!

**Comment by DaleMckeown** [comment]
I just came across an issue being caused by long PR names, and was surprised to find that there is no way to control how PR names are formatted.

My Dependabot target-branch is `dependabot/main`. When Dependabot creates a PR, it is created using the following format:

`dependabot/{packageManager}/{targetBranchName}/{packageName}`. In my case, this becomes `dependabot/nuget/dependabot/main/{packageName}`. 

I'd love the flexibility to be able to omit the `target-branch` value from being included in the PR name.

This would allow by Dependabot PRs to be: `dependabot/nuget/{packageName}`, as I originally intended.

**Comment by pukanszkypeter-x** [comment]
👍🏼 
 
Yeah, it would be nice if there would be an option that would allow you to enter custom branch names in the `pull-request-branch-name` option and not just the separators.

Alternatively, you could look at Renovate's insane array of configuration options 😄: 
[https://docs.renovatebot.com/configuration-options/](https://docs.renovatebot.com/configuration-options/)

Please, it's opened since 2018 😢

**Comment by Geigerkind** [comment]
We use cloud tools that fail because of long branch names. This makes using dependabot really annoying. 
I would really love if this quick win could get a tiny little bit more love :(

**Comment by victor-ferrer-form3** [comment]
Had to resort to this in my Github Workflow file:

```
# Sanitizes dependabot branch names
echo "suffix=-$HEADREF" | sed -r 's/[-.]+/_/g' >> $GITHUB_OUTPUT
```

**Comment by jeffwidman** [comment]
> We use cloud tools that fail because of long branch names. This makes using dependabot really annoying.

The length is already configurable:
* https://github.com/dependabot/dependabot-core/pull/5338

And we're talking about adding some reasonable per-platform defaults too:
* https://github.com/dependabot/dependabot-core/issues/7564

What is not captured in those is modifying the _format_ of branch names, which is much more complex then just truncating to a max length. That's what this ticket is about.

**Comment by jamie-phlo** [comment]
> > We use cloud tools that fail because of long branch names. This makes using dependabot really annoying.
> 
> The length is already configurable:
> 
> * [Add max length option to BranchNamer #5338](https://github.com/dependabot/dependabot-core/pull/5338)
> 
> And we're talking about adding some reasonable per-platform defaults too:
> 
> * [Set defaults for the max length of PR branch names #7564](https://github.com/dependabot/dependabot-core/issues/7564)
> 
> What is not captured in those is modifying the _format_ of branch names, which is much more complex then just truncating to a max length. That's what this ticket is about.

Although the max length is configurable it doesn't seem exposed in the GitHub actions config options, https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file

How feasible is it to add this in? Is there a repo for the GitHub action, I'd be happy to try and contribute towards this but unsure if the action code is in a private repo or one that doesn't accept contributions from the community.

**Comment by mhanci92** [comment]
Also waiting, and really looking forward for exposure of the branch length configurability so that in can be set in the `dependabot.yml`, as the branch name normally provides the necessary uniqueness for our Helm releases. As known, Helm & Kubernetes restrict the length for names that we can give resources to be created. We currently need ugly workarounds in order to make our Helm releases work in our automated CI/CD setup, in which we build the application with the version change provided by the dependabot MR (as we work with GitLab), deploy it to the Kubernetes cluster and run automated e2e-tests on it.

**Comment by austinpray-mixpanel** [comment]
In our primary repo if a branch name exceeds 85 chars in our environment we run into 

>The size of mapped attribute exceeds the 127 bytes limit.

Authenticating to GCP using workload identity federation via https://github.com/google-github-actions/auth

see: https://github.com/google-github-actions/auth/blob/87b651ab4a60ba1164458f539d64a58d055f4b03/docs/TROUBLESHOOTING.md#subject-exceeds-the-127-byte-limit

dependabot violates this length limit pretty often
![image](https://github.com/dependabot/dependabot-core/assets/71290498/8a09d7ec-9d3e-4559-8d9d-26003e8eaec3)

**Comment by erzz** [comment]
Also hit by max length issues with workload identity and the subject being longer than 127 bytes.  We cannot rely on Dependabot when there are limitations like this and will revert back to Renovate which is FAR more configurable.

Two specific options you can simply steal I would say are:

```
"branchPrefix": "rv8-",
"hashedBranchLength": 20,
```

All branch names are a simple hash like `rv8-11381c5613945835`

Its already super apparent from the PR and the files changed what dependencies are being updated - not sure why we would want all that detail in a branch name too

**Comment by tophercullen** [comment]
@felixbouleau @victor-ferrer-form3 There appears to be an option to configure the separator: 
https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#pull-request-branch-nameseparator

**Comment by kirrg001** [comment]
This breaks on Tekton pipelines. The separator is not enough.

> PersistentVolumeClaim "pvc-dependabot-npm_and_yarn-main-nan-2.19.0-a6f7692d-71b1-4b96-917d-af3baad2b1c0-artifacts" is invalid: metadata.name: Invalid value: "pvc-dependabot-npm_and_yarn-main-nan-2.19.0-a6f7692d-71b1-4b96-917d-af3baad2b1c0-artifacts": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')

**Comment by ashbork** [comment]
Is there any progress on this, at least when it comes to a customisable branch name length?

**Comment by austinpray-mixpanel** [comment]
My org started paying GitHub a ton of money for GitHub Advanced Security. Maybe I can convince someone at GitHub to TAL at this now.

The branch name length issue is legitimately blocking us using dependabot for modules nested in our monorepo.

<img width="889" alt="Image" src="https://github.com/user-attachments/assets/5b246348-9e20-4ceb-aea6-36b339e8a3ac" />

**Comment by DuyTran-TomTom** [comment]
We still don't have the `max-length` option in Github dependabot? Does no one bother to look?

**Comment by erzz** [comment]
Kinda crazy that 7 years on .... there is zero progress on this?

**Comment by ctufekci** [comment]
Is there any progress ?

**Comment by baoqchau** [comment]
It's 8 years since this issue was opened. No updates at all?

**Comment by AbhishekBhaskar** [comment]
Hi everyone! 👋

We're excited to announce that **configurable branch name formatting** is now available under `pull-request-branch-name` in your `dependabot.yml`. This addresses the most common pain points raised in this thread: Docker tag compatibility, ACR naming requirements, branch length limits, and general formatting flexibility.

## New Options

The following new options are available under `pull-request-branch-name` for each update entry:

| Option | Description | Example Values |
|--------|-------------|----------------|
| `separator` | Replaces `/` between branch name segments | `"-"`, `"_"`, `"/"` (default) |
| `prefix` | Custom prefix (replaces `dependabot`) | `"deps"`, `"my-project"` |
| `max-length` | Maximum branch name length (20–244) | `63`, `128` |
| `word-separator` | Replaces `_` in branch content (e.g. `npm_and_yarn` → `npm-and-yarn`) | `"-"`, `"."` |
| `branch-name-case` | Transform branch content to lowercase or uppercase | `"lowercase"`, `"uppercase"` |
| `template` | Full control over branch name format using placeholders | `"{prefix}/{package_manager}/{dependency}-{version}"` |

All options are **composable** — you can combine `template` with `word-separator`, `branch-name-case`, etc.

## Examples

### Docker / ACR compatible (no slashes, no underscores, lowercase)


```yaml
pull-request-branch-name:
  separator: "-"
  word-separator: "-"
  branch-name-case: "lowercase"
```

**Before:** `dependabot/npm_and_yarn/Lodash-4.17.21`
**After:** `dependabot-npm-and-yarn-lodash-4.17.21`

### Branch length limits (Kubernetes, Tekton, monorepo paths)

```yaml
pull-request-branch-name:
  separator: "-"
  word-separator: "-"
  branch-name-case: "lowercase"
  max-length: 63
```

When the branch name exceeds `max-length`, it's truncated with a hash suffix to preserve uniqueness.

### Custom prefix

```yaml
pull-request-branch-name:
  prefix: "deps"
```

**Result:** `deps/npm_and_yarn/lodash-4.17.21`

### Template (full control)

For advanced use cases where you need complete control over the branch format:

```yaml
pull-request-branch-name:
  template: "{prefix}/{package_manager}/{dependency}-{version}"
```

**Available placeholders:**

| Placeholder | Description |
|---|---|
| `{prefix}` | Branch name prefix (default `"dependabot"`) |
| `{package_manager}` | e.g. `npm_and_yarn`, `pip`, `docker` |
| `{directory}` | Dependency file directory |
| `{target_branch}` | Target branch if configured |
| `{dependency}` | Dependency name (solo updates only) |
| `{version}` | New version (solo updates only) |
| `{group_name}` | Group name (grouped updates only) |
| `{name}` | Auto-selects: dependency+version for solo, group name for groups |

Templates work with grouped updates too — for dependency groups and multi-ecosystem groups, a content digest is automatically appended to guarantee branch uniqueness.

### Combining template with formatting options

Template and simple options are composable — formatting is applied as post-processing after template rendering:

```yaml
pull-request-branch-name:
  template: "{prefix}/{package_manager}/{dependency}-{version}"
  separator: "-"
  word-separator: "-"
  branch-name-case: "lowercase"
  max-length: 80
```

**Result:** `dependabot-npm-and-yarn-lodash-4.17.21`

## Full example `dependabot.yml`

```yaml
version: 2

multi-ecosystem-groups:
  infrastructure:
    schedule:
      interval: "weekly"
    pull-request-branch-name:
      template: "{prefix}/infra/{name}"
      word-separator: "-"
      branch-name-case: "lowercase"

updates:
  - package-ecosystem: "npm"
    directory: "/"
    schedule:
      interval: "weekly"
    pull-request-branch-name:
      separator: "-"
      word-separator: "-"
      branch-name-case: "lowercase"

  - package-ecosystem: "docker"
    directory: "/"
    schedule:
      interval: "monthly"
    pull-request-branch-name:
      template: "{prefix}/{package_manager}/{dependency}-{version}"
      max-length: 60

  - package-ecosystem: "pip"
    directory: "/backend"
    schedule:
      interval: "weekly"
    pull-request-branch-name:
      prefix: "deps"
      branch-name-case: "lowercase"

  # These entries participate in the "infrastructure" multi-ecosystem group
  - package-ecosystem: "docker"
    directory: "/infra"
    patterns: ["nginx", "redis", "postgres"]
    multi-ecosystem-group: "infrastructure"

  - package-ecosystem: "terraform"
    directory: "/infra"
    patterns: ["hashicorp/*"]
    multi-ecosystem-group: "infrastructure"
```

**Resulting branch names:**

| Scenario | Strategy | Branch |
|---|---|---|
| lodash npm update | Solo | `dependabot-npm-and-yarn-lodash-4.17.21` |
| nginx solo docker update | Solo | `dependabot/docker/nginx-1.25.0` |
| pip Django update | Solo | `deps/pip/django-4.2.1` |
| nginx solo in /infra (not grouped) | Solo | `dependabot-docker-nginx-1.25.0` |
| Cross-eco infrastructure group (docker + terraform) | MultiEcosystem | `dependabot/infra/infrastructure-fc93691fd4` |

**Key points for multi-ecosystem groups:**
- The `pull-request-branch-name` on `multi-ecosystem-groups.infrastructure` controls the **grouped cross-ecosystem PR** branch name.
- Individual `updates` entries that specify `multi-ecosystem-group` **cannot** have their own `pull-request-branch-name` — the group-level configuration takes precedence and is the only one used for those entries.
- `{package_manager}` is **not available** in multi-ecosystem group templates (would fail validation since it spans multiple ecosystems).
- A content digest is always auto-appended to multi-ecosystem group branches to guarantee uniqueness.

## Important notes

- **Config is per-update entry** — each `package-ecosystem` entry can have its own branch name configuration.
- **Existing PRs are not affected** — changes only apply to newly created PRs. Close/merge existing PRs for them to be recreated with the new format.
- **Default behavior is unchanged** — if you don't configure any options, branch names remain exactly as they are today.
- **`separator`** (the existing option) continues to work as before and can be combined with all new options.

We'd love to hear your feedback! Please let us know if you run into any issues/bugs. Thank you!

**Comment by AbhishekBhaskar** [comment]
Closing this as completed.
